[Ader] 사다리 구현 4단계 - 리팩토링 2

ak2j38 opened the pull request:
## 4단계 미션 PR입니다!
항상 코드리뷰해주셔서 감사합니다 호눅스!
- - -
### 구현사항
- [x] 미션3 코드리뷰 적용
- [x] 사다리의 각 위치요소를 Element와 Line으로 관리
    - [x] Element 클래스 추가
    - [x] Line 클래스 추가
    - [x] 발판 생성 여부를 위의 두 클래스에서 확인
- [x] 유효성 검사 패키지 및 클래스 추가
- [x] 테스트 코드 추가
    - [x] 이름 유효성 검사 테스트
    - [x] 발판 생성 로직 테스트

### 고민사항
1. Line의 생성자에서 발판의 생성 로직이 수행되는데 이 경우에는 Line을 하나씩 생성해가면서 테스트를 진행해야 할까요? 발판생성 로직 자체는 private메소드입니다.

ak2j38 committed:
docs : README 미션 4 내용 추가

1. README에 미션 4내용 추가

end

ak2j38 committed:
feature : 미션3 코드리뷰 내용 수정

1. InputView에서 플레이어 이름의 최대치인 5를 매직넘버로 쓰고 있던 것 상수로 변경

end

ak2j38 committed:
feature : 사다리의 각 위치 요소에 해당하는 Element 클래스 추가

1. Element 클래스는 index와 value(세로선, 발판, 빈칸), 발판여부를 갖고 있습니다.

end

ak2j38 committed:
feature : RandomUtils 클래스 생성

1. Ladder 클래스의 책임을 줄이고 추후 추가될 수 있는 랜덤값들을 위해 클래스 분리

end

ak2j38 committed:
feature : Line을 통해 사다리의 각 줄을 관리

1. Element클래스의 생성자 private에서 public으로 변경
2. Line 클래스 생성

end

ak2j38 committed:
feature : 패키지 depth 1 추가

1. 각 패키지를 관리하는 Ladder 클래스를 추가했습니다.

end

ak2j38 committed:
feature : 사다리를 Line과 Element를 통해 관리하도록 수정

1. 각 줄의 요소를 담당하는 Element 클래스 추가
2. Ladder 클래스의 책임을 줄이기 위해 클래스 분할 및 생성자 변경

end

ak2j38 committed:
test : 이전 칸에 발판유무와 랜덤값의 유무에 따라 현재 발판 여부 테스트 추가

1. 이전 칸에 발판이 있으면 다음 칸엔 발판이 생성되지 않는다
2. 이전 칸에 발판이 없고 랜덤값이 참이라면 다음 칸엔 발판이 생성된다
3. 이전 칸에 발판이 없고 랜덤값이 거짓이라면 다음 칸엔 발판이 생성되지 않는다

end

ak2j38 committed:
feature : 유효성검사 클래스 추가 및 메소드 분리

1. 입력 이름값 5글자 검증을 테스트하기 위해 패키지 및 메소드 분리

end

ak2j38 committed:
feature : 이름 유효성 검사 이름 길이 최소조건 추가

1. 입력 이름값은 최소 1글자 이상이어야한다

end

ak2j38 committed:
test : 이름 유효성 검사 테스트 추가

1. 입력 이름값은 최소 1글자 이상이며 5글자 이하를 검사하는 테스트 추가

end

ak2j38 committed:
docs : 미션4 문서 완성

1. 미션4 내용 추가

end

tmdgusya commented on src/main/java/Ladder/domain/Line.java:
```diff
@@ -0,0 +1,51 @@
+public class Line {
+
+    private static final String BLANK_FOR_PRINT = "  ";
+    private List<Element> elements;
```
Line 은 elements 가 존재해야만 동작하는 일급 콜렉션이 아닐까요?
위와 같은 이유로 한번 다시 생각해보시는걸 추천드립니다.

ak2j38 replied:
남겨주신 리뷰 덕분에 일급컬렉션에 대해서 학습했습니다! 

불필요하게 존재하던 멤버변수를 메소드내에 인스턴스 변수로 이동시켰습니다.
추가로 Element 리스트를 final 키워드로 변경했습니다.

tmdgusya commented on src/main/java/Ladder/domain/Element.java:
```diff
@@ -0,0 +1,40 @@
+        if (index % 2 == 0) {
+            return LADDER_VERTICAL_LINE;
+        }
+        if (index % 2 != 0 && !hasPrevStep && rand) {
```
위에서 이미 `index % 2 == 0` 에 대한 검증을 치룬거 같은데요.
굳이 밑에서 `index % 2 != 0` 을 검증안해주어도 될거 같다는 생각이 드네요.
아미 26번 라인까지 왔다면 이미 저 케이스는 통과했을거란 생각이 듭니다.

ak2j38 replied:
맞습니다 ㅠㅠ 인텔리제이에서도 이미 노란블럭으로 알려주고 있었네요! 감사합니다 🙇

tmdgusya commented on src/main/java/Ladder/validation/Validator.java:
```diff
@@ -0,0 +1,21 @@
+
+public class Validator {
+
+    private static final String PLAYER_NAME_DELIMITER = ",";
```
변수명 좋네요 👍

tmdgusya commented on src/test/java/Ladder/validation/ValidatorTest.java:
```diff
@@ -0,0 +1,35 @@
+class ValidatorTest {
+
+    @Test
+    @DisplayName("사용자가 입력한 이름들이 최소 1자리 이상 최대 5자리를 만족할 때")
```
사용자가 입력한 이름들이 최소 1자리 이상 최대 5자리를 만족할때 어떤 결과가 나온다는 것일까요?
저는 @DisplayName 은 보통 `~ 일때 ~ 이다` 라고 하는 편입니다. 
if ... then 식으로 주어진 상황과 결과에 대한 이야기를 적어보시는건 어떨까요?

ak2j38 replied:
다른사람이 볼때는 True를 의도하는지 False를 의도하는지 알 수 없군요!
가이드 주신대로 "~일 때 ~ 이다" 형태로 변경해보겠습니다!

tmdgusya commented on src/test/java/Ladder/validation/ValidatorTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    @DisplayName("사용자가 입력한 이름들이 최소 1자리 이상 최대 5자리를 만족할 때")
+    void isNameMaxLengthFiveTest1() {
+        // given
```
given.. when.. then 사용 좋네요 👍

tmdgusya commented on src/test/java/Ladder/domain/ElementTest.java:
```diff
@@ -0,0 +1,59 @@
+        element4 = new Element(3, false, true);    // expected : -----
+
+        // then
+        assertTrue(element4.getHasStep());
```
굳이 getter 라고해서 getXXX 식의 naming 을 가져갈 이유는 없습니다 :)
getHasStep 이 나은지 hasStep 이 나은지 한번 고민해보시죠.

더 좋게 표현할 수 있을까요?

ak2j38 replied:
자바빈 규약을 무의식적으로 사용하다보니 메소드 네이밍이 무언가 이상해진 것 같네요 ㅠㅠ 

1. 기존 멤버변수로 가지고 있던 boolean형 변수인 hasStep과 메소드명이 중복되는 것 같아 변수를 제거하고 hasStep() 메소드 하나만 유지하는 것으로 변경예정입니다.
2. 추가적으로 Element 클래스에 불필요한 멤버변수들을 정리 예정입니다.

tmdgusya commented on src/main/java/Ladder/domain/Line.java:
```diff
@@ -0,0 +1,51 @@
+        return sb.toString();
+    }
+
+    private void initElementList() {
```
ElementList 의 시작이라는 뜻으로 해석되는데..
ElementList 를 만들어주는 메소드가 더 맞지 않을까요? 메소드명을 한번 더 고민해보시길 바랍니다.

ak2j38 replied:
메소드명을 고민하던 중 불필요한 메소드라고 판단되어서 삭제했습니다 😀

tmdgusya commented on src/main/java/Ladder/domain/Ladder.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Ladder {
```
전반적으로 이 클래스에 대해서는 고려할 점이 많은데요.
일단 일급 콜렉션에 대한 내용을 전반적으로 학습해보시는걸 추천드립니다.
https://jojoldu.tistory.com/412

1. 일단 이 클래스의 코드를 보았을때 인스턴스화가 되는 동시에 반드시 ladderInstance 가 필요하다고 생각됩니다.
그렇다면 List 앞에 어떤 키워드를 붙이는게 좋을까요? `final` 에 대해서 한번 공부해보시면 좋을거 같아요.

2. getLadder 가 의미가 있는 클래스인지 한번 생각해보시는것 또한 추천드립니다.
제 생각엔 저게 있으나 없으나 다를게 없다고 생각되네요.
```java
 public Ladder(int playersCount, int maxLadderHeight) {
        this.ladderInstance = createLadder(playersCount, maxLadderHeight);
    }
```
로도 충분해 보이기도 해서요.

3. 생성자를 를 통해서 ladderinstance 를 생성하고 있는데요.
일단 ladderinstance 라는 변수명또한 맞는지 생각해 봐야할 필요가 있습니다.
instance 라는 단어의 개념또한 정리해볼 필요가 있다고 생각됩니다.

ak2j38 replied:
첨부해주신 링크와 검색을 통해 일급컬렉션에 대해 학습해보았습니다. 🙇🙇🙇

Ladder 클래스 같은 경우에는 초반 미션을 통해 싱글톤 패턴을 적용해서 구현해보았는데 이후 미션들을 통해 변경되며 애매해진 것 같습니다 ㅠㅠ
아래 사항들을 통해 개선해보겠습니다!

1.  확실히 이전에는 final을 상수 문자열에만 기계적으로 사용하던 느낌이었는데 이런 경우에도 사용해야한다는 것을 새로 학습했습니다!
2. 싱글톤 패턴으로 구현때 만들었던 메소드인데 아직 남아있네요... 생성자에서 한번에 생성할 수 있도록 개선해보겠습니다.
3. 이것또한 이펙티브 자바를 통해 싱글톤을 흉내내다보니 아무생각없이 instance를 뒤에 붙인 것 같습니다. 더 좋은 네이밍을 생각해보겠습니다.

tmdgusya reviewed (changes requested):
안녕하세요 리뷰어 로치입니다 :)

전반적으로 변수명에 대해서는 깔끔한 부분이 많아서 좋았습니다
Delimiter 나 is 를 사용한 부분은 boolean 으로 사용한다 등 이해하기 쉽게 코드를 짜신 부분이 많았습니다 👍 
어느정도 잘 작성하신거 같아 학습할 부분과 고쳐보면 좋을거 같은 부분에 대한 가이드를 드립니다.
다시 한번 수정해서 올리면 좋을거 같아서 request change 드립니다.

ak2j38 committed:
feature : 코드리뷰 내용 개선

1. Line 클래스를 일급컬렉션으로 변경(Element 리스트만 가지고 있음)
2. Element 클래스내에 불필요한 조건 제거
3. Element 클래스에서 관리되던 전역멤버변수 hasStep 삭제(hasStep() 메소드로 대체) 및 미사용 멤버변수 index 삭제
4. 테스트 클래스 @DisplayName "~일 때 ~이다" 형식으로 변경
5. Line 클래스의 initElementList() 불필요하다고 판단하여 메소드 삭제
6. Ladder 클래스의 불필요 메소드를 정리하고 일급컬렉션의 형태로 개선

end

ak2j38 commented:
리뷰 남겨주신 내용을 바탕으로 코드 개선을 해보았습니다.

전체적으로 일급컬렉션에 대해 학습하고 적용해보았으며 불필한 메소드들을 정리했습니다.
또한 테스트코드에 대한 DisplayName도 다시한번 생각해볼 수 있었습니다.

좋은 리뷰를 남겨주셔서 감사합니다 로치 😁😁😁

tmdgusya commented:
아 혹시 수정한 버전에 대해서 다시 push 해서 보내주실수 있으실까요?
그러면 merge 시켜드리겠습니다 :)
전반적으로 코드가 깔끔해서 리뷰할게 많이 없었는데 수정하신 부분들이 궁금하네요 ㅎㅎ
4단계 미션 하시느라 고생많으셨습니다 💯

ak2j38 commented:
> 아 혹시 수정한 버전에 대해서 다시 push 해서 보내주실수 있으실까요? 그러면 merge 시켜드리겠습니다 :) 전반적으로 코드가 깔끔해서 리뷰할게 많이 없었는데 수정하신 부분들이 궁금하네요 ㅎㅎ 4단계 미션 하시느라 고생많으셨습니다 💯

@tmdgusya 
엇... 제가 오늘 날짜로 커밋했는데 혹시 확인 불가능하신건가요....? 
65f4855969ad557040d436d8675959074998204e     <--- 2월 19일 커밋입니다!
